File: constructs/go/compute.go
package infra

import (
	"github.com/aws/constructs-go/constructs/v10"
	"github.com/aws/jsii-runtime-go"
	"github.com/hashicorp/terraform-cdk-go/cdktf"
	"github.com/k1cka5h/infra-go/policy"
)

const aksModuleSource = "git::ssh://git@github.com/k1cka5h/terraform-modules.git" +
	"//modules/compute/aks?ref=v1.4.0"

// NodePoolConfig defines an additional AKS node pool.
type NodePoolConfig struct {
	VMSize     *string
	NodeCount  *float64
	MinCount   *float64
	MaxCount   *float64
	Taints     *[]*string
	Labels     *map[string]*string
}

// AksConfig configures the AKS cluster.
type AksConfig struct {
	KubernetesVersion   *string
	SystemNodeCount     *float64
	SystemVMSize        *string
	AdditionalNodePools *map[string]*NodePoolConfig
}

// AksConstructProps configures an AksConstruct.
type AksConstructProps struct {
	Project       *string
	Environment   *string
	ResourceGroup *string
	Location      *string
	SubnetId      *string
	LogWorkspaceId *string
	Aks           *AksConfig
	ExtraTags     *map[string]*string
}

// AksConstruct provisions an AKS cluster with Azure CNI and AAD RBAC.
// Wraps modules/compute/aks from the platform Terraform module repo.
type AksConstruct struct {
	constructs.Construct
	module cdktf.TerraformModule
}

// NewAksConstruct creates and returns an AksConstruct.
func NewAksConstruct(scope constructs.Construct, id *string, props *AksConstructProps) *AksConstruct {
	c := &AksConstruct{Construct: constructs.NewConstruct(scope, id)}

	aks := props.Aks
	if aks == nil {
		aks = &AksConfig{}
	}

	k8sVersion := jsii.String("1.29")
	if aks.KubernetesVersion != nil {
		k8sVersion = aks.KubernetesVersion
	}

	systemCount := jsii.Number(3)
	if aks.SystemNodeCount != nil {
		systemCount = aks.SystemNodeCount
	}

	systemVMSize := jsii.String("Standard_D4s_v3")
	if aks.SystemVMSize != nil {
		systemVMSize = aks.SystemVMSize
	}

	additionalPools := map[string]interface{}{}
	if aks.AdditionalNodePools != nil {
		for name, pool := range *aks.AdditionalNodePools {
			p := map[string]interface{}{
				"vm_size":    pool.VMSize,
				"node_count": pool.NodeCount,
				"min_count":  pool.MinCount,
				"max_count":  pool.MaxCount,
			}
			if pool.Taints != nil {
				p["node_taints"] = pool.Taints
			}
			if pool.Labels != nil {
				p["node_labels"] = pool.Labels
			}
			additionalPools[name] = p
		}
	}

	c.module = cdktf.NewTerraformModule(c, jsii.String("aks"), &cdktf.TerraformModuleConfig{
		Source: jsii.String(aksModuleSource),
		Variables: &map[string]interface{}{
			"project":               props.Project,
			"environment":          props.Environment,
			"resource_group_name":  props.ResourceGroup,
			"location":             props.Location,
			"subnet_id":            props.SubnetId,
			"log_analytics_workspace_id": props.LogWorkspaceId,
			"kubernetes_version":   k8sVersion,
			"system_node_count":    systemCount,
			"system_vm_size":       systemVMSize,
			"additional_node_pools": additionalPools,
			"tags":                 policy.RequiredTags(props.Project, props.Environment, props.ExtraTags),
		},
	})

	return c
}

func (a *AksConstruct) ClusterId()                *string { return a.module.GetString(jsii.String("cluster_id")) }
func (a *AksConstruct) KubeletIdentityObjectId()  *string { return a.module.GetString(jsii.String("kubelet_identity_object_id")) }

File: constructs/typescript/src/compute.ts
import { Construct } from "constructs";
import { TerraformModule } from "cdktf";
import { requiredTags } from "./policy/tagging";

const MODULE_SOURCE =
  "git::ssh://git@github.com/nautilus/terraform-modules.git" +
  "//modules/compute/aks?ref=v1.4.0";

export interface NodePoolConfig {
  readonly vmSize?:            string;
  readonly nodeCount?:         number;
  readonly enableAutoScaling?: boolean;
  readonly minCount?:          number;
  readonly maxCount?:          number;
  readonly labels?:            Record<string, string>;
  readonly taints?:            string[];
}

export interface AksConfig {
  readonly kubernetesVersion?:     string;
  readonly systemNodeVmSize?:      string;
  readonly systemNodeCount?:       number;
  readonly additionalNodePools?:   Record<string, NodePoolConfig>;
  readonly adminGroupObjectIds?:   string[];
  readonly serviceCidr?:           string;
  readonly dnsServiceIp?:          string;
  readonly extraTags?:             Record<string, string>;
}

export interface AksConstructProps {
  readonly project:        string;
  readonly environment:    string;
  readonly resourceGroup:  string;
  readonly location:       string;
  readonly subnetId:       string;
  readonly logWorkspaceId: string;
  readonly config?:        AksConfig;
}

export class AksConstruct extends Construct {
  private readonly _module: TerraformModule;

  constructor(scope: Construct, id: string, props: AksConstructProps) {
    super(scope, id);

    const cfg = props.config ?? {};

    const additionalNodePools: Record<string, unknown> = {};
    for (const [name, pool] of Object.entries(cfg.additionalNodePools ?? {})) {
      additionalNodePools[name] = {
        vm_size:             pool.vmSize            ?? "Standard_D4s_v3",
        node_count:          pool.nodeCount         ?? 2,
        enable_auto_scaling: pool.enableAutoScaling ?? false,
        min_count:           pool.minCount          ?? 1,
        max_count:           pool.maxCount          ?? 10,
        labels:              pool.labels            ?? {},
        taints:              pool.taints            ?? [],
      };
    }

    this._module = new TerraformModule(this, "aks", {
      source: MODULE_SOURCE,
      variables: {
        project:                    props.project,
        environment:                props.environment,
        resource_group_name:        props.resourceGroup,
        location:                   props.location,
        subnet_id:                  props.subnetId,
        log_analytics_workspace_id: props.logWorkspaceId,
        kubernetes_version:         cfg.kubernetesVersion   ?? "1.29",
        system_node_vm_size:        cfg.systemNodeVmSize    ?? "Standard_D2s_v3",
        system_node_count:          cfg.systemNodeCount     ?? 3,
        additional_node_pools:      additionalNodePools,
        admin_group_object_ids:     cfg.adminGroupObjectIds ?? [],
        service_cidr:               cfg.serviceCidr         ?? "10.240.0.0/16",
        dns_service_ip:             cfg.dnsServiceIp        ?? "10.240.0.10",
        tags: requiredTags(props.project, props.environment, cfg.extraTags),
      },
    });
  }

  get clusterId(): string                    { return this._module.getString("cluster_id"); }
  get clusterName(): string                  { return this._module.getString("cluster_name"); }
  /** Assign ACR pull and Key Vault read to this identity. */
  get kubeletIdentityObjectId(): string      { return this._module.getString("kubelet_identity_object_id"); }
  /** Assign Network Contributor to this identity. */
  get clusterIdentityPrincipalId(): string   { return this._module.getString("cluster_identity_principal_id"); }
}
